ansible_base/resource_registry/tasks/sync.py

```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
+    #   incredibly difficult to pull off.
+
+    conflict_resource = resource_type.get_conflicting_resource(resource_data)
+
+    if conflict_resource is None:
+        return
+
+    resp = api_client.get_resource(conflict_resource.ansible_id)
+
+    # If the conflicting resource doesn't exist on the server, just go ahead and delete it.
+    # This will most likely happen with resources that are ignored by get_orphan_resources.
+    if resp.status_code == 404:
+        delete_resource(conflict_resource)
+
+    # If the resource does exist, lets update it first. Hopefully this desn't also result
+    # in a duplicate key error. If it does, we're cooked.
+    elif resp.status_code == 200:
+        data = resp.json()
+        conflict_resource.update_resource(
+            resource_data=data["resource_data"], service_id=data["service_id"], is_partially_migrated=data["is_partially_migrated"]
+        )
+    else:
+        raise ResourceSyncHTTPError
+
+
 def _attempt_update_resource(
     manifest_item: ManifestItem,
     resource: Resource,
     resource_data: dict,
+    api_client: ResourceAPIClient,
     **kwargs,
 ) -> SyncResult:
     """Try to update existing resource."""
     try:
         resource.update_resource(resource_data, partial=True, **kwargs)
     except IntegrityError:  # pragma: no cover
-        return SyncResult(SyncStatus.CONFLICT, manifest_item)
+        # This typically means that there was a duplicate key error. To mitigate this
+        # we will attempt to hanlde the conflicting resource and perform the operation
+        # again.
+        try:
+            _handle_conflict(resource_data, resource.resource_type_obj, api_client)
+            resource.update_resource(resource_data, partial=True, **kwargs)
+        except (ResourceDeletionError, IntegrityError) as e:
+            logger.error(f"Failed to gracefully handle conflict for {resource_data}. Got error {e}.")
+            return SyncResult(SyncStatus.CONFLICT, manifest_item)
     except Error as e:
+        # Something happened with the database. We don't know what it is. Instead of failing the whole
+        # sync, we'll raise an error and skip this for now.
         logger.error(f"Failed to update resource {resource.ansible_id}. Received error: {e}")
         return SyncResult(SyncStatus.ERROR, manifest_item)
-    else:
-        return SyncResult(SyncStatus.UPDATED, manifest_item)
+
+    return SyncResult(SyncStatus.UPDATED, manifest_item)
+
+
+def _attempt_create_resource(
+    manifest_item: ManifestItem,
+    resource_data: dict,
+    resource_type: ResourceType,
+    resource_service_id: str,
+    api_client: ResourceAPIClient,
+) -> SyncResult:
+    try:
+        Resource.create_resource(
+            resource_type=resource_type,
+            resource_data=resource_data,
+            ansible_id=manifest_item.ansible_id,
+            service_id=resource_service_id,
+        )
```

Comment:
The pattern of try and try again is duplicated. It would be really sweet if we could figure out a way of abstracting this so we could work on the "try again" logic in one place instead of two.

Reply:
I thought about making this function recursive, but that got complicated very quickly. This repeats itself, but it's much more understandable.